### PR TITLE
[IMP] survey: small changes over the whole module

### PR DESCRIPTION
**Pull request by wde-odoo**
Changing the naming of fields, moving actions, buttons and deleting obsolete actions
as well as changing the front-end to improve the UI.

Some fields were mislabelled or unclear, menus were only available within
developer mode and the front-end needed to be polished.

task_id=2600241

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/survey/models/survey_question.py [hunk 1]
```diff
@@ -553,6 +553,9 @@ class SurveyQuestionAnswer(models.Model):
     value_image = fields.Image('Image', max_width=256, max_height=256)
     is_correct = fields.Boolean('Is a correct answer')
     answer_score = fields.Float('Score for this choice', help="A positive score indicates a correct choice; a negative or null score indicates a wrong answer")
+    question_type = fields.Selection(related='question_id.question_type')
```
Comment (awa-odoo) [inline addons/survey/models/survey_question.py:556]: I would move those under the "question_id" field and create a small section with a python comment.

Something like:

```py
# question and question related fields
question_id = fields.Many2one('survey.question', string='Question', ondelete='cascade')
question_type = fields.Selection(related='question_id.question_type')
scoring_type = fields.Selection(related='question_id.scoring_type')
allow_value_image = fields.Boolean(related='question_id.allow_value_image')
# answer specific fields
...
```

It's usually good to give some structure when we start having a lot of fields.

File: addons/survey/models/survey_survey.py [hunk 2]
```diff
@@ -19,6 +19,7 @@ class Survey(models.Model):
     _description = 'Survey'
     _rec_name = 'title'
     _inherit = ['mail.thread', 'mail.activity.mixin']
+    _order = "id desc"
```
Comment (awa-odoo) [inline addons/survey/models/survey_survey.py:22]: Specification says:

> Instead, let's order Survey by Create Date (newest one first)

So shouldn't it be "create_date desc" instead?

File: addons/survey/static/src/js/fields_section_one2many.js [hunk 3]
```diff
@@ -51,7 +51,11 @@ var SectionListRenderer = ListRenderer.extend({
                 if (record.data.questions_selection === "random") {
                     nbrColumns--;
                 }
+                // Render empty cells for buttons to avoid having unaligned elements
```
Comment (awa-odoo) [inline addons/survey/static/src/js/fields_section_one2many.js:54]: I remember we discussed this change but I'm not sure what it's about anymore.
Can you explain again?

File: addons/survey/views/survey_question_views.xml [hunk 4]
```diff
@@ -29,42 +28,50 @@
                             <field name="question_type" widget="radio" attrs="{'required': [('is_page', '=', False)]}" />
                         </group>
                         <group>
-                            <div class="col-lg-6 offset-lg-3 o_preview_questions">
+                            <div class="col-lg-6 offset-lg-3 o_preview_questions" style="width:100%;height:101px">
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:31]: A few things here:

- We try to avoid custom styling in form as much as possible.
- If it's not possible to avoid id, we try to use bootstrap classes as much as possible (e.g here: "w-100")
- In any case, you should (almost) never write style attributes, use scss instead
- 101px height is oddly specific... what are you trying to do here?

File: addons/survey/views/survey_question_views.xml [hunk 5]
```diff
@@ -29,42 +28,50 @@
...
-                                    <i class="fa fa-2x" role="img" aria-label="Numeric" title="Numeric">123..</i>
+                                    <i>How many ...?</i><br/>
+                                    <i class="fa fa-2x" role="img" aria-label="Numeric" title="Numeric">123&#160;</i>
+                                    <i class="fa fa-2x fa-sort" role="img" aria-label="Numeric"/>
                                 </div>
                                 <!-- Date -->
                                 <div attrs="{'invisible': [('question_type', '!=', 'date')]}">
-                                    <p class="o_datetime">YYYY-MM-DD
+                                    <i>When is Christmas?</i><br/>
+                                    <p class="o_datetime" style="border:none">YYYY-MM-DD
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:51]: Same remark here, replace the style with a "border-0" class.

File: addons/survey/views/survey_question_views.xml [hunk 6]
```diff
@@ -29,42 +28,50 @@
...
+                                    <div class="row"><i class="fa fa-dot-circle-o fa-lg"/> &#160;&#160;answer</div>
+                                    <div class="row"><i class="fa fa-circle-o  fa-lg"/> &#160;&#160;answer</div>
                                 </div>
                                 <!-- Multiple choice: multiple answers allowed -->
                                 <div attrs="{'invisible': [('question_type', '!=', 'multiple_choice')]}" role="img" aria-label="Multiple choice with multiple answers" title="Multiple choice with multiple answers">
-                                    <div class="row"><i class="fa fa-square-o fa-lg"/> answer</div>
-                                    <div class="row"><i class="fa fa-check-square-o fa-lg"/> answer</div>
-                                    <div class="row"><i class="fa fa-square-o fa-lg"/> answer</div>
+                                    <i>Which are yellow?</i><br/>
+                                    <div class="row"><i class="fa fa-square-o fa-lg"/> &#160;&#160;answer</div>
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:72]: I can see you use "&#160;" for spacing (I'm guessing).

Please use spacing classes instead (e.g: "mr-1", see bootstrap doc for more details).
Apply for other occurrences of "&#160;" too.

File: addons/survey/views/survey_question_views.xml [hunk 7]
```diff
@@ -148,7 +155,7 @@
                                         <field name="sequence" widget="handle"/>
                                         <field name="value" string="Choices"/>
                                         <field name="value_image" options="{'accepted_file_extensions': 'image/*'}"
-                                            attrs="{'column_invisible': ['|', ('parent.allow_value_image', '=', False), ('parent.question_type', '=', 'matrix')]}"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:151]: Not sure why this change?

File: addons/survey/views/survey_question_views.xml [hunk 8]
```diff
@@ -267,6 +273,34 @@
             </tree>
         </field>
     </record>
+
+    <record id="survey_question_answer_form" model="ir.ui.view">
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:277]: Should be "survey_question_answer_view_form"

File: addons/survey/views/survey_question_views.xml [hunk 9]
```diff
@@ -267,6 +273,34 @@
...
+        <field name="name">survey.question.answer.view.form</field>
+        <field name="model">survey.question.answer</field>
+        <field name="arch" type="xml">
+            <form string="question answer form">
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:281]: We usually capitalize those ("Question Answer Form")

File: addons/survey/views/survey_survey_views.xml [hunk 10]
```diff
@@ -9,6 +9,7 @@
                 <field name="id" invisible="1"/>
                 <field name="session_state" invisible="1"/>
                 <header>
+                    <field name="question_ids" invisible="1"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_survey_views.xml:12]: Can you put this just above the action that uses it?

File: addons/survey/views/survey_survey_views.xml [hunk 11]
```diff
@@ -83,68 +84,68 @@
...
+                                        <field name="certification" nolabel="1"/>
+                                        <div attrs="{'invisible': [('certification', '=', False)]}">
+                                            <field name="certification_report_layout" class="oe_inline"/>
+                                            <button name="action_survey_preview_certification_template"
+                                                string="Preview" type="object"
+                                                icon="fa-external-link"  target="_blank" class="btn-link pt-0"/>
+                                        </div>
+                                    </div> 
+                                    <field name="certification_mail_template_id"/>    
+                                    <div colspan="2" class="o_checkbox_optional_field" attrs="{'invisible': ['|', ('certification', '=', False), ('users_login_required', '=', False)]}">
```
Comment (awa-odoo) [inline addons/survey/views/survey_survey_views.xml:129]: I think this messes a bit the layout:

image

See the vertical bar is not aligned correctly.
Can you double-check?

File: addons/survey/views/survey_survey_views.xml [hunk 12]
```diff
@@ -281,6 +289,9 @@
                     domain="[('my_activity_date_deadline', '=', context_today().strftime('%Y-%m-%d'))]"/>
                 <filter invisible="1" string="Upcoming Activities" name="activities_upcoming_all"
                     domain="[('my_activity_date_deadline', '&gt;', context_today().strftime('%Y-%m-%d'))]"/>
+                <group expand="0" string="Group By">
+                    <filter name="gfilterroup_by_user_id" string="User Id" domain="[]" context="{'group_by': 'user_id'}"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_survey_views.xml:293]: "gfilterroup_by_user_id" ? :smile:

File: addons/survey/views/survey_templates.xml [hunk 13]
```diff
@@ -137,7 +137,8 @@
                 <t t-else="">
                     <div t-if="survey.is_time_limited">
                         <p>
-                            <span>Time limit for this survey: </span>
+                            <span t-if="not survey.certification">Time limit for this survey: </span>
+                            <span t-if="survey.certification">Time limit for this certification: </span>
```
Comment (awa-odoo) [inline addons/survey/views/survey_templates.xml:141]: Could use "t-if / t-else" notation instead.

File: addons/survey/views/survey_templates.xml [hunk 14]
```diff
@@ -466,6 +471,7 @@
             <t t-set="item_idx" t-value="0"/>
             <div class="d-flex flex-wrap col-lg-12">
                 <t t-set="has_correct_answer" t-value="scoring_display_correction and any(label.is_correct for label in question.suggested_answer_ids)"/>
+                <t t-set="n_answ" t-value="len(question.suggested_answer_ids)"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_templates.xml:474]: Naming: "answers_count" ?

File: addons/survey/views/survey_templates.xml [hunk 15]
```diff
@@ -476,20 +482,22 @@
...
-                               t-att-checked="'checked' if answer_line else None"
-                               t-att-data-selection-key="letters[item_idx] if useKeySelection else ''"/>
-                        <span class="ml-2" t-field='label.value'/>
-                        <i class="fa fa-check-square float-right mt-1 position-relative"></i>
-                        <i class="fa fa-square-o float-right mt-1 position-relative"></i>
-                        <t t-call="survey.question_suggested_value_image"/>
-                    </label>
+                    <div t-att-class="'%s pl-0 pr-2' % ('col-6' if n_answ in [1, 2, 4] else 'col-4')">
+                         <label t-att-class="'col-12 o_survey_choice_btn mr-2 mb-2 py-3 pl-4 pr-5 rounded %s %s' % (answer_class, 'o_survey_selected' if answer_line else '')"
+                                style="height: calc(100% - 0.5rem); line-break: auto; line-break: auto;">
```
Comment (awa-odoo) [inline addons/survey/views/survey_templates.xml:487]: - Should use scss
- Please add a small comment in the scss file about the calc height thing
- Looks like you have twice "line-break"

(Same for the other template).

File: addons/survey/views/survey_templates.xml [hunk 16]
```diff
@@ -401,10 +402,11 @@
         <div class="row o_survey_form_choice"
              t-att-data-name="question.id"
              data-question-type="simple_choice_radio">
-            <t t-set="item_idx" t-value="0"/>
             <div t-attf-class="col-lg-12 d-flex flex-wrap">
                 <t t-set="has_correct_answer" t-value="scoring_display_correction and any(label.is_correct for label in question.suggested_answer_ids)"/>
+                <t t-set="n_answ" t-value="len(question.suggested_answer_ids)"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_templates.xml:407]: Naming: "answers_count" ?

File: addons/survey/views/survey_templates.xml [hunk 17]
```diff
@@ -401,10 +402,11 @@
...
                 <t t-foreach='question.suggested_answer_ids' t-as='label'>
+                    <t t-set="answer_line" t-value="answer_lines.filtered(lambda line: line.suggested_answer_id)"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_templates.xml:409]: Looks like this is set above? (line 399/400)

**Review by awa-odoo — COMMENTED** [review]
Hello @wde-odoo 

As discussed, we would like to speed this one up so here is a first validation pass.

Lots of small comments, which is to be expected with a task with a lot of small tweaks.
One bigger note about the commit message, you can find details in the comments.

Don't hesitate to ping me on Discord if anything is unclear.

Cheers!

File: addons/survey/static/src/js/fields_section_one2many.js (hunk 3 above)
Comment (wde-odoo) [reply addons/survey/static/src/js/fields_section_one2many.js:54]: It's for https://www.awesomescreenshot.com/image/11096552?key=127083eb310eefe2beaa8971caad44b0

When we activate the "Randomized per Section" option, there was a problem with the alignment because the section took too much space. By reducing the number of columns the section takes, we can create enough space pour the "Random Question Count" column the be properly alligned.

File: addons/survey/views/survey_question_views.xml (hunk 7 above)
Comment (wde-odoo) [reply addons/survey/views/survey_question_views.xml:151]: Because of this:

LNA CHANGE REQUEST 09/09/2021:
    Let's change a bit the "image" implementation:
    Instead of having it as an option here: https://tinyurl.com/yfaxoty3
    We always display the image field https://tinyurl.com/ygbvo75k
=> And we adapt the display of the survey as soon as an image is uploaded for this question

File: addons/survey/views/survey_question_views.xml (hunk 4 above)
Comment (wde-odoo) [reply addons/survey/views/survey_question_views.xml:31]: I don't know why i used 101px. I changed it to 100px.

File: addons/survey/views/survey_survey_views.xml (hunk 11 above)
Comment (wde-odoo) [reply addons/survey/views/survey_survey_views.xml:129]: I don't seem to have that problem. Do you still have it?
Here's what I got
image

File: addons/survey/controllers/main.py [hunk 18]
```diff
@@ -191,7 +191,7 @@ def _prepare_survey_finished_values(self, survey, answer, token=False):
         values = {'survey': survey, 'answer': answer}
         if token:
             values['token'] = token
-        if survey.scoring_type != 'no_scoring' and survey.certification:
```
Comment (awa-odoo) [inline addons/survey/controllers/main.py:194]: Is this after the testing?
Maybe worth a bullet point in your commit message.

File: addons/survey/models/survey_question.py [hunk 19]
```diff
@@ -546,12 +546,18 @@ class SurveyQuestionAnswer(models.Model):
...
     sequence = fields.Integer('Label Sequence order', default=10)
+    scoring_type = fields.Selection(related='question_id.scoring_type')
+    allow_value_image = fields.Boolean(related='question_id.allow_value_image')
+    matrix_question_id = fields.Many2one('survey.question', string='Question (as matrix row)', ondelete='cascade')
+
+    #answer related fields
     value = fields.Char('Suggested value', translate=True, required=True)
-    value_image = fields.Image('Image', max_width=256, max_height=256)
     is_correct = fields.Boolean('Is a correct answer')
+    value_image = fields.Image('Image', max_width=256, max_height=256)
```
Comment (awa-odoo) [inline addons/survey/models/survey_question.py:560]: No need to move this I'd say.

File: addons/survey/static/src/scss/survey_questions_views.scss [hunk 20]
```diff
@@ -0,0 +1,4 @@
+.o_preview_questions {
```
Comment (awa-odoo) [inline addons/survey/static/src/scss/survey_questions_views.scss:1]: When playing around on the form view on F12 and disabling those rules, I don't see much change.
Is this to add some padding simply?

Can't you use padding classes instead? (e.g: "p-1" / "p-2" / ...)?

File: addons/survey/views/survey_question_views.xml (hunk 7 above)
Comment (awa-odoo) [reply addons/survey/views/survey_question_views.xml:151]: Ok, but what about "allow_value_image" then?
Is it even displayed anywhere anymore?
If we don't need it, we need to clean it and remove it.

File: addons/survey/views/survey_user_views.xml [hunk 21]
```diff
@@ -152,7 +154,7 @@
         <field name="name">survey.user_input.line.view.form</field>
         <field name="model">survey.user_input.line</field>
         <field name="arch" type="xml">
-            <form string="User input line details" create="false">
+            <form string="User input line details" create="false" edit="false">
```
Comment (awa-odoo) [inline addons/survey/views/survey_user_views.xml:157]: Why removing the edition on those? If the user has the rights?

**Review by awa-odoo — COMMENTED** [review]
Hello @wde-odoo 

Small counter-validation after your first pass.

Cheers!

**Review by awa-odoo — COMMENTED** [review]
Hello @wde-odoo 

A couple more comments.
I think we are ready to send this into testing though so we can already make some iterations on that side.

Thanks!

File: addons/survey/static/src/scss/survey_survey_views.scss [hunk 22]
```diff
@@ -24,3 +24,8 @@ table.o_section_list_view tr.o_data_row.o_is_section {
 .icon_rotates {
     transform: rotate(180deg);
 }
+
+.o_preview_questions {
+    width: 100%;
```
[the remark this answers is not in the document]
Comment (wde-odoo) [reply addons/survey/static/src/scss/survey_survey_views.scss:29]: I just saw that the width is already set at 100% in web/static/src/legacy/scss/import_bootstrap.scss

File: addons/survey/views/survey_question_views.xml (hunk 7 above)
Comment (wde-odoo) [reply addons/survey/views/survey_question_views.xml:151]: I skipped it by accident. It is now done.

File: addons/survey/models/survey_survey.py [hunk 23]
```diff
@@ -317,6 +318,11 @@ def _compute_certification_give_badge(self):
                not survey.certification:
                 survey.certification_give_badge = False
 
+    @api.depends('certification')
+    def _compute_scoring_type(self):
+        if self.scoring_type == 'no_scoring':
```
Comment (awa-odoo) [inline addons/survey/models/survey_survey.py:323]: Missing a "certification" check probably?

File: addons/survey/views/survey_user_views.xml (hunk 21 above)
Comment (awa-odoo) [reply addons/survey/views/survey_user_views.xml:157]: Not sure why this comment is resolved?
Maybe we discussed it and I don't remember.

File: addons/survey/views/survey_question_views.xml [hunk 24]
```diff
@@ -97,9 +103,9 @@
                         </group>
                     </group>
                     <notebook>
-                        <page string="Answers" name="answers" attrs="{'invisible': [('is_page', '=', True)]}">
+                        <page string="Answers" name="answers" attrs="{'invisible': ['|', ('is_page', '=', True), ('question_type', '=', 'text_box')]}">
                             <div class="row">
-                                <div class="col-6"
+                                <div class="col-6"
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:108]: Extra end-of-line space here.
(There are some IDE extensions that can help with that)

**Review by awa-odoo — COMMENTED** [review]
Hello @wde-odoo 

Has a quick morning check, looks good overall except a strange compute.
Can you also double-check your commit message to avoid that extra spacing?
image

Thanks & cheers!

File: addons/survey/models/survey_question.py [hunk 25]
```diff
@@ -49,7 +49,7 @@ class SurveyQuestion(models.Model):
     def default_get(self, fields):
         defaults = super(SurveyQuestion, self).default_get(fields)
         if (not fields or 'question_type' in fields):
-            defaults['question_type'] = False if defaults.get('is_page') == True else 'text_box'
+            defaults['question_type'] = False if defaults.get('is_page') == True else 'simple_choice'
```
Comment (awa-odoo) [inline addons/survey/models/survey_question.py:52]: Runbot ci/style is complaining about this one.
Could remove the "== True" I think.

File: addons/survey/views/survey_question_views.xml [hunk 26]
```diff
@@ -142,15 +148,22 @@
                                     </group>
                                 </div>
                             </div>
-                            <group attrs="{'invisible': [('question_type', 'not in', ['simple_choice', 'multiple_choice', 'matrix'])]}">
+                            <group attrs="{'invisible': [('question_type', 'not in', ['simple_choice', 'multiple_choice'])]}">
+                                <field name="suggested_answer_ids" context="{'default_question_id': active_id}" nolabel="1">
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:152]: Also, why do we have the field "suggested_answer_ids" twice now?
It's usually not a good idea to have the same field twice in a view.
Wasn't it not working properly with the column_invisible checks?

File: addons/survey/views/survey_question_views.xml [hunk 27]
```diff
@@ -167,40 +180,41 @@
...
-                            <group string="Constraints">
-                                <group>
-                                    <field name="constr_mandatory" string="Mandatory Answer"/>
-                                    <field name="constr_error_msg" attrs="{'invisible': [('constr_mandatory', '=', False)]}"/>
+                            <group>
+                                <group string="Answers">
+                                    <field name="save_as_email" string="Input must be an email"/>
+                                    <field name="save_as_nickname"/>
+                                    <field name="validation_required"/>
+                                    <field name="column_nb" string="Number of columns" attrs="{'invisible':[('question_type','not in',['simple_choice', 'multiple_choice'])]}"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_question_views.xml:188]: We usually keep one space between every arguments.

So 
`{'invisible':[('question_type','not in',['simple_choice', 'multiple_choice'])]}`

Becomes

`{'invisible':[('question_type', 'not in', ['simple_choice', 'multiple_choice'])]}`

File: addons/survey/views/survey_templates.xml [hunk 28]
```diff
@@ -406,7 +407,7 @@
         <div class="row o_survey_form_choice"
              t-att-data-name="question.id"
              data-question-type="simple_choice_radio">
-            <t t-set="item_idx" t-value="0"/>
+             <t t-set="item_idx" t-value="0"/>
```
Comment (awa-odoo) [inline addons/survey/views/survey_templates.xml:410]: Why the extra space?

File: addons/survey/views/survey_user_views.xml (hunk 21 above)
Comment (wde-odoo) [reply addons/survey/views/survey_user_views.xml:157]: Sorry, thought I replied.

It's to prevent changing answers on a survey and messing with the results.

File: addons/survey/views/survey_question_views.xml [hunk 29]
```diff
@@ -29,43 +27,51 @@
                             <field name="question_type" widget="radio" attrs="{'required': [('is_page', '=', False)]}" />
                         </group>
                         <group>
-                            <div class="col-lg-6 offset-lg-3 o_preview_questions">
+                            <div class="col-lg-6 offset-lg-3 w-100 o_preview_questions">
+                                <!-- Multiple choice: only one answer -->
+                                <div attrs="{'invisible': [('question_type', '!=', 'simple_choice')]}" role="img" aria-label="Multiple choice with one answer" title="Multiple choice with one answer">
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_question_views.xml:32]: Linting (for this whole PR): when having long lines, you may define as multi-line to ease reading, like
```
<div attrs="{'invisible': [('question_type', '!=', 'simple_choice')]}" role="img"
     aria-label="Multiple choice with one answer"
     title="Multiple choice with one answer">
    <i>...</i>
</div>
```

File: addons/survey/models/survey_survey.py [hunk 30]
```diff
@@ -20,6 +20,7 @@ class Survey(models.Model):
     _order = 'create_date DESC'
     _rec_name = 'title'
     _inherit = ['mail.thread', 'mail.activity.mixin']
+    _order = "create_date desc"
```
Comment (tde-banana-odoo) [inline addons/survey/models/survey_survey.py:23]: Seems already done 4 lines above ^^

File: addons/survey/models/survey_survey.py [hunk 31]
```diff
@@ -87,23 +88,23 @@ def _get_default_session_code(self):
...
-    success_ratio = fields.Integer("Success Ratio", compute="_compute_survey_statistic")
+    success_ratio = fields.Integer("Success Ratio %", compute="_compute_survey_statistic")
     # scoring
     scoring_type = fields.Selection([
         ('no_scoring', 'No scoring'),
         ('scoring_with_answers', 'Scoring with answers at the end'),
         ('scoring_without_answers', 'Scoring without answers at the end')],
-        string="Scoring", required=True, default='no_scoring')
-    scoring_success_min = fields.Float('Success %', default=80.0)
+        string="Scoring", required=True, default='no_scoring', store=True, readonly=False, compute="_compute_scoring_type")
```
Comment (tde-banana-odoo) [inline addons/survey/models/survey_survey.py:106]: I think you can multi-line the definition, beginning to get out of my screen ^^

File: addons/survey/static/src/scss/survey_survey_views.scss [hunk 32]
```diff
@@ -173,3 +173,13 @@
         }
     }
 }
+
+.o_preview_questions {
+    min-height: 100px;
+}
+
+.o_survey_question_view_form {
```
Comment (tde-banana-odoo) [inline addons/survey/static/src/scss/survey_survey_views.scss:181]: To be moved in a ``survey_question_views.scss`` (this file is for surveys only normally)

File: addons/survey/static/src/scss/survey_survey_views.scss [hunk 33]
```diff
@@ -173,3 +173,13 @@
         }
     }
 }
+
+.o_preview_questions {
```
Comment (tde-banana-odoo) [inline addons/survey/static/src/scss/survey_survey_views.scss:177]: Should be under ``o_survey_question_view_form``

File: addons/survey/views/survey_question_views.xml [hunk 34]
```diff
@@ -16,9 +16,7 @@
                     <div class="oe_title" style="width: 100%;">
                         <label for="title" string="Section" attrs="{'invisible': [('is_page', '=', False)]}"/>
                         <label for="title" string="Question" attrs="{'invisible': [('is_page', '=', True)]}"/>
-                        <separator />
-                        <field name="title" colspan="4"/>
-                        <separator />
+                        <field name="title" placeholder="e.g. &quot;What is the...&quot;" colspan="4"/>
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_question_views.xml:19]: A bit strange but ok, seems it is difficult to have a really dynamic placeholder ^^ .

File: addons/survey/views/survey_question_views.xml [hunk 35]
```diff
@@ -143,19 +149,21 @@
                                 </div>
                             </div>
                             <group attrs="{'invisible': [('question_type', 'not in', ['simple_choice', 'multiple_choice', 'matrix'])]}">
-                                <field name="suggested_answer_ids" string="Answers" context="{'default_question_id': active_id}">
+                                <field name="suggested_answer_ids" context="{'default_question_id': active_id}" nolabel="1">
                                     <tree editable="bottom">
                                         <field name="sequence" widget="handle"/>
-                                        <field name="value" string="Choices"/>
+                                        <field name="value" string="Choices" style="width: 50%"/>
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_question_views.xml:155]: mmh, why a hardcoded styling ? Seems to have a lot of "width: 50%" ...

File: addons/survey/views/survey_question_views.xml [hunk 36]
```diff
@@ -303,5 +341,10 @@
         action="survey_question_answer_action"
         parent="survey_menu_questions"
         sequence="3"/>
+        <menuitem name="Detailed Answers"
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_question_views.xml:344]: Linting: small indentation issue here at menuitem definition

File: addons/survey/views/survey_survey_views.xml [hunk 37]
```diff
@@ -83,68 +84,68 @@
...
+                                        <div attrs="{'invisible': [('certification_give_badge', '=', False)]}">
+                                            <field name="certification_badge_id"
+                                                attrs="{'invisible': ['|', ('certification_give_badge', '=', False), ('certification_badge_id', '!=', False)], 'required': [('certification_give_badge', '=', True)]}"
+                                                domain="[('survey_id', '=', active_id), ('survey_id', '!=', False)]"
+                                                context="{'default_name': title,
+                                                        'default_description': 'Congratulations, you have succeeded this certification',
+                                                        'default_rule_auth': 'nobody',
+                                                        'default_level': None,
+                                                        'form_view_ref': 'survey.gamification_badge_form_view_simplified',
+                                                        'default_website_published': True}"/>
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_survey_views.xml:141]: Why would be publish it by default ? Especially this is valid only in website_profile module ... would remove it, unless you really want it ?

Edit: ok, this is a move .. let us keep it then, nobody complained.

File: addons/survey/views/survey_survey_views.xml [hunk 38]
```diff
@@ -83,68 +84,68 @@
...
+                                    <field name="certification_mail_template_id"/>
+                                    <div colspan="2" class="float-left o_checkbox_optional_field" attrs="{'invisible': ['|', ('certification', '=', False), ('users_login_required', '=', False)]}">
+                                        <label for="certification_give_badge"/>
+                                        <field name="certification_give_badge"/>
+                                        <div attrs="{'invisible': [('certification_give_badge', '=', False)]}">
+                                            <field name="certification_badge_id"
+                                                attrs="{'invisible': ['|', ('certification_give_badge', '=', False), ('certification_badge_id', '!=', False)], 'required': [('certification_give_badge', '=', True)]}"
+                                                domain="[('survey_id', '=', active_id), ('survey_id', '!=', False)]"
+                                                context="{'default_name': title,
+                                                        'default_description': 'Congratulations, you have succeeded this certification',
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_survey_views.xml:137]: This will not be translated, which is not good. Somehow you will have to define a real action I fear if you want to go at that level of details.

Edit: ok, this is a move .. let us keep it then, nobody complained.

File: addons/survey/views/survey_survey_views.xml [hunk 39]
```diff
@@ -18,10 +18,11 @@
                         attrs="{'invisible': [('session_state', '=', False)]}" />
                     <button name="action_end_session" string="Close Live Session" type="object"
                         attrs="{'invisible': [('session_state', 'not in', ['ready', 'in_progress'])]}" />
-                    <button name="action_test_survey" string="Test" type="object" attrs="{'invisible': [('active', '=', False)]}"/>
-                    <button name="action_print_survey" string="Print" type="object"/>
+                    <field name="question_ids" invisible="1"/>
```
Comment (tde-banana-odoo) [inline addons/survey/views/survey_survey_views.xml:21]: Would maybe move it with other question fields (or maybe before header), even if invisible let us order them a bit nicely :) Header is for button, not something else.

**Review by tde-banana-odoo — COMMENTED** [review]
Review ! Thanks for your PR, added some comments while reading it :) . Feel free to ping if anything is unclear. Keep faith !

File: addons/survey/views/survey_survey_views.xml (hunk 39 above)
Comment (wde-odoo) [reply addons/survey/views/survey_survey_views.xml:21]: I was told it is good practice to put the invisible field right before it is used... Should I still move it up?

File: addons/survey/data/mail_template_data.xml [hunk 40]
```diff
@@ -62,10 +67,11 @@
             <tr><td style="padding:15px 20px 10px 20px;">
                 <p>Dear <span t-out="object.partner_id.name or 'participant'">participant</span></p>
                 <p>
-                    Here is, in attachment, your certification document for
-                        <strong t-out="object.survey_id.display_name or ''">Feedback Form</strong>
+                    Please find attached your
+                        <strong t-out="object.survey_id.display_name or ''"/>
```
[the remark this answers is not in the document]
Comment (wde-odoo) [reply addons/survey/data/mail_template_data.xml:71]: I'm not sure what you mean with "demo-filler-data". I also might have misunderstood what was asked.
I did this change because of:

Current:
Dear Billy Fox
Here is, in attachment, your certification document for Furniture Creation Certification
Congratulations for succeeding the test!

To Be
Dear Billy Fox,
Please find attached your SurveyName certification.
Congratulations for passing the test!

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/survey/data/mail_template_data.xml:71]: Actually in
```
Here is, in attachment, your certification document for
<strong t-out="object.survey_id.display_name or ''">Feedback Form</strong>
```
Feedback Form is a filler in case there is no object (for a no-record rendering), allowing to see "how it looks like in real life". You should therefore keep it, like
```
Please find attached your
<strong t-out="object.survey_id.display_name or ''">Furniture Creation</strong>
certification
```
(or anything better than Furniture Creation, used as a demo data / filler data)

File: addons/survey/tests/test_survey_security.py [hunk 41]
```diff
@@ -256,19 +256,18 @@ def test_access_answers_survey_user(self):
...
         answers = self.env['survey.user_input'].search([('survey_id', 'in', [survey_own.id, self.survey.id])])
         self.assertEqual(answers, answer_own | self.answer_0)
 
         answer_lines = self.env['survey.user_input.line'].search([('survey_id', 'in', [survey_own.id, self.survey.id])])
-        self.assertEqual(answer_lines, answer_line_own | self.answer_0_0 | self.answer_0_1)
 
         self.env['survey.user_input'].browse(answer_own.ids).read(['state'])
         self.env['survey.user_input'].browse(self.answer_0.ids).read(['state'])
 
-        self.env['survey.user_input.line'].browse(answer_line_own.ids).read(['value_numerical_box'])
```
Comment (tde-banana-odoo) [inline addons/survey/tests/test_survey_security.py:271]: Why just removing ? Go until the end of your test and update everything accordingly (probably a assertRaises I suppose).

**Review by tde-banana-odoo — COMMENTED** [review]
Review: models / acls

File: addons/survey/security/ir.model.access.csv [hunk 42]
```diff
@@ -17,7 +17,7 @@ access_survey_user_input_survey_user,survey.user_input.survey.user,model_survey_
 access_survey_user_input_survey_manager,survey.user_input.survey.manager,model_survey_user_input,group_survey_manager,1,1,1,1
 access_survey_user_input_line_all,survey.user_input.line.all,model_survey_user_input_line,,0,0,0,0
 access_survey_user_input_line_user,survey.user_input.line.user,model_survey_user_input_line,base.group_user,0,0,0,0
-access_survey_user_input_line_survey_user,survey.user_input.line.survey.user,model_survey_user_input_line,group_survey_user,1,1,1,1
+access_survey_user_input_line_survey_user,survey.user_input.line.survey.user,model_survey_user_input_line,group_survey_user,1,0,0,0
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/survey/security/ir.model.access.csv:20]: Edit: actually still having read access so should not cause issues

**Comment by tde-banana-odoo** [comment]
@robodoo r+
